In [1]:
with open("test1.txt", "rt") as f:
    test1 = f.readlines()

In [17]:
with open("test2.txt", "rt") as f:
    test2 = f.readlines()

In [1]:
with open("input", "rt") as f:
    inp = f.readlines()

In [35]:
def part1(text: list[str]) -> int:
    dirs = text[0].strip()
    desert = {}
    for l in text[2:]:
        desert.update({l[0:3]: (l[7:10], l[12:15])})
    steps = 0
    pos = 'AAA'
    while 1:
        if dirs[steps%len(dirs)] == 'L':
            pos = desert[pos][0]
        else:
            pos = desert[pos][1]
        steps += 1
        if pos == 'ZZZ':
            return steps

In [36]:
part1(test1)

2

In [37]:
part1(test2)

6

In [39]:
with open("output1", "wt") as f:
    f.write(str(part1(inp)))

## Part 2

In [37]:
dirs = inp[0].strip()
desert = {}
for l in inp[2:]:
        desert.update({l[0:3]: (l[7:10], l[12:15])})

starts = [k for k in desert.keys() if k[-1] == 'A']

In [4]:
len(dirs)

271

In [6]:
starts

['LPA', 'XKA', 'QGA', 'AAA', 'HHA', 'LTA']

In [24]:
for s in starts:
    pos = s
    d = {}
    for i in range(1, 65000):
        if dirs[i%len(dirs)] == 'L':
            pos = desert[pos][0]
        else:
            pos = desert[pos][1]
        if pos[-1] == 'Z':
            if pos in d.keys():
                d[pos].append(i)
            else:
                d[pos] = [i,]
    print(d)

{'BRZ': [12194, 30351, 48508]}
{'QSZ': [2709, 14362, 26015, 37668, 49321, 60974]}
{'MCZ': [1896, 23305, 44714]}
{'ZZZ': [4606, 17343, 30080, 42817, 55554]}
{'LRZ': [2438, 16801, 31164, 45527, 59890]}
{'BKZ': [7316, 23305, 39294, 55283]}


* Every ghost only visits one Z-node
* Periodic
* Every frequency is divisible by 271 (the frequency of the direction tape),
* and also the first occurrence of Z-node,
* so we van work just by whole tapes without skipping any Z-node

In [70]:
p0 = []
freq = []

for s in starts:
    pos = s
    d = []
    for i in range(1, 300):
        for j in range(len(dirs)):
            if dirs[j] == 'L':
                pos = desert[pos][0]
            else:
                pos = desert[pos][1]
        if pos[-1] == 'Z':
            d.append(i)
    print(f'{d[0]}\t{d[1]-d[0]}')
    p0.append(d[0])
    freq.append(d[1]-d[0])

67	67
43	43
79	79
47	47
53	53
59	59


In [41]:
freq

[67, 43, 79, 47, 53, 59]

All the Z-nodes are found after an exact period (no Chinese Remainder Theorem), and all the frequencies are primes, so their LCM is their product:

In [73]:
p = len(dirs)
for f in freq:
    p *= f
print(p)

9064949303801
